sgl-kernel/csrc/moe/ep_moe_silu_and_mul_kernel.cu

```diff
@@ -0,0 +1,107 @@
+#include <ATen/ATen.h>
+#include <ATen/cuda/CUDAContext.h>
+#include <c10/cuda/CUDAGuard.h>
+#include <cuda_bf16.h>
+#include <cuda_fp16.h>
+
+#include <THC/THCAtomics.cuh>
+#include <flashinfer/vec_dtypes.cuh>
+
+#include "utils.h"
+
+using namespace flashinfer;
+
+template <typename scalar_t>
+__device__ inline scalar_t silu_quantize(float x);
+
+template <>
+__device__ inline float silu_quantize<float>(float x) {
+  float y = x / (1.f + __expf(-x));
+  return y;
+}
+
+template <>
+__device__ inline __half silu_quantize<__half>(float x) {
+  float y = x / (1.f + __expf(-x));
+  return __float2half_rn(y);
+}
+
+template <>
+__device__ inline __nv_bfloat16 silu_quantize<__nv_bfloat16>(float x) {
+  float y = x / (1.f + __expf(-x));
+  return __float2bfloat16_rn(y);
+}
+
+template <typename scalar_t>
+__global__ void ep_moe_act_and_mul_cuda_kernel(
+    const scalar_t* __restrict__ gateup_output,
+    scalar_t* __restrict__ down_input,
+    const int* __restrict__ reorder_topk_ids,
+    const float* __restrict__ scales,
+    int start_expert_id,
+    int end_expert_id,
+    int hidden_size) {
+  constexpr uint32_t vec_size = 16 / sizeof(scalar_t);
+  using vec_t = flashinfer::vec_t<scalar_t, vec_size>;
+
+  const int64_t token_idx = blockIdx.x;
+  const int64_t thread_idx = threadIdx.x;
+  const int64_t stride = blockDim.x;
+
+  const int half_hidden_size = hidden_size >> 1;
+  const int expert_id = reorder_topk_ids[token_idx];
+
+  if (expert_id < start_expert_id || expert_id > end_expert_id) return;
+  const scalar_t* gate_output_ptr = gateup_output + static_cast<int64_t>(token_idx) * hidden_size;
+  const scalar_t* up_output_ptr = gate_output_ptr + half_hidden_size;
+  scalar_t* dst_ptr = down_input + static_cast<int64_t>(token_idx) * half_hidden_size;
+  scalar_t scale_q = static_cast<scalar_t>(scales ? (1.f / scales[expert_id - start_expert_id]) : 1.f);
+
+  const uint32_t vec_elements = half_hidden_size / vec_size;
```

Comment:
It might be worth considering how to handle the tail elements when hidden_size is not divisible by vec_size.